### PR TITLE
Add timestamps to the OutputManager's info_map data when adding logs, errors, and warnings.

### LINKED ISSUE #598: Add timestamp to logs, warnings, and errors in the Output Manager
**Is your feature request related to a problem? Please describe.**
We don't have a way to follow and track the sequence of events.

**Describe the solution you'd like**
In functions `add_log`, `add_warning`, and `add_error`, add a timestamp entry in the `info_map' before calling the `_add_to_pool`.

**Describe alternatives you've considered**

1. We could move the timestamp adding logic to `_add_to_pool` and avoid repetition, but that will add timestamps for variables as well. I think that would be unnecessary and TMI.
2. We could let the callers of  `add_log`, `add_warning`, and `add_error` to add their timestamps if they want. But that would lead to too many changes across the codebase. 

**Additional context**

### PR DESCRIPTION
**Pull request by michaelRichards99**
Added timestamp info to the OutputManager when adding logs, errors, and warnings. Consolidated file timestamp creation to use the same method. Added checks to the existing unit tests to check that the timestamp information is present.

Fixes #598.

### CODE REVIEW

File: RUFAS/output_manager.py [hunk 1]
```diff
@@ -187,9 +189,21 @@ def add_error(self, name: str, msg: str, info_map: Dict[str, Any]) -> None:
         info_map["suffix"] : str, optional
             If present, gets appended to the key
         """
+        self._add_timestamp_to_info_map(info_map)
         key = self._generate_key(name, info_map)
         self._add_to_pool(self.errors_pool, key, msg, info_map)
 
+    def _get_timestamp(self, include_millis: bool = False) -> str:
+        """Produce the current system time as a timestamp string."""
+        base_timestamp_str: str = "%d-%b-%Y_%a_%H-%M-%S"
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py:217]: Can you add an example in the docstring to show what this looks like?

File: RUFAS/output_manager.py [hunk 2]
```diff
@@ -187,9 +189,21 @@ def add_error(self, name: str, msg: str, info_map: Dict[str, Any]) -> None:
         info_map["suffix"] : str, optional
             If present, gets appended to the key
         """
+        self._add_timestamp_to_info_map(info_map)
         key = self._generate_key(name, info_map)
         self._add_to_pool(self.errors_pool, key, msg, info_map)
 
+    def _get_timestamp(self, include_millis: bool = False) -> str:
+        """Produce the current system time as a timestamp string."""
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
        """Produces the current system time as a timestamp string."""
```

File: RUFAS/output_manager.py [hunk 3]
```diff
@@ -187,9 +189,21 @@ def add_error(self, name: str, msg: str, info_map: Dict[str, Any]) -> None:
...
+        base_timestamp_str: str = "%d-%b-%Y_%a_%H-%M-%S"
+        timestamp_format_string: str = f"{base_timestamp_str}.%f" if include_millis else base_timestamp_str
+        return datetime.datetime.now().strftime(timestamp_format_string)
+
+    def _add_timestamp_to_info_map(self, info_map: Dict[str, Union[str, bool]]) -> None:
+        """Add the current system time to the info_map."""
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
        """Adds the current system time to the info_map."""
```

File: RUFAS/output_manager.py [hunk 4]
```diff
@@ -187,9 +189,21 @@ def add_error(self, name: str, msg: str, info_map: Dict[str, Any]) -> None:
...
         key = self._generate_key(name, info_map)
         self._add_to_pool(self.errors_pool, key, msg, info_map)
 
+    def _get_timestamp(self, include_millis: bool = False) -> str:
+        """Produce the current system time as a timestamp string."""
+        base_timestamp_str: str = "%d-%b-%Y_%a_%H-%M-%S"
+        timestamp_format_string: str = f"{base_timestamp_str}.%f" if include_millis else base_timestamp_str
+        return datetime.datetime.now().strftime(timestamp_format_string)
+
+    def _add_timestamp_to_info_map(self, info_map: Dict[str, Union[str, bool]]) -> None:
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
    def _add_timestamp_to_info_map(self, info_map: Dict[str, Any]) -> None:
```
Why Union[str, bool]?

File: RUFAS/output_manager.py [hunk 5]
```diff
@@ -310,7 +324,7 @@ def _generate_file_name(self, base_name: str, extension: str = "json") -> str:
         """
         Returns a file name using the given base_name and timestamp.
         """
-        timestamp = time.strftime(r"%d-%b-%Y_%a_%H-%M-%S", time.localtime())
+        timestamp: str = self._get_timestamp()
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
        timestamp: str = self._get_timestamp(include_millis=False)
```

File: tests/test_misc.py [hunk 6]
```diff
@@ -368,6 +368,8 @@ def test_add_error(
     mock_output_manager.add_error(name, value, info_map)
 
     mock_output_manager._generate_key.assert_called_once_with(name, info_map)
+
+    assert info_map["timestamp"] is not None
```
Comment (PooyaHekmati) [inline tests/test_misc.py]: Let's also check for the correctness of the value. You'd need to mock datetime.datetime.now() and manually set its return value.

File: tests/test_misc.py [hunk 7]
```diff
@@ -395,6 +396,8 @@ def test_add_warning(
     mock_output_manager.add_warning(name, value, info_map)
 
     mock_output_manager._generate_key.assert_called_once_with(name, info_map)
+
+    assert info_map["timestamp"] is not None
```
Comment (PooyaHekmati) [inline tests/test_misc.py]: Same

File: tests/test_misc.py [hunk 8]
```diff
@@ -422,6 +424,8 @@ def test_add_log(
     mock_output_manager.add_log(name, value, info_map)
 
     mock_output_manager._generate_key.assert_called_once_with(name, info_map)
+
+    assert info_map["timestamp"] is not None
```
Comment (PooyaHekmati) [inline tests/test_misc.py]: Same

File: RUFAS/output_manager.py (hunk 4 above)
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]: Maybe this function is overkill and not needed?

File: RUFAS/output_manager.py [hunk 9]
```diff
@@ -160,6 +161,7 @@ def add_warning(self, name: str, msg: str, info_map: Dict[str, Any]) -> None:
         info_map["suffix"] : str, optional
             If present, gets appended to the key
         """
+        self._add_timestamp_to_info_map(info_map)
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
        info_map["timestamp"] = self._get_timestamp(include_millis=True)
```

**Review by PooyaHekmati — CHANGES_REQUESTED** [review]
Nice work! It's good to have functions with clear responsibilities. The newly developed functions need to have test cases as well.

File: RUFAS/output_manager.py (hunk 9 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: done

File: RUFAS/output_manager.py (hunk 4 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: yeah it's not really necessary. removed.

File: tests/test_misc.py (hunk 8 above)
Comment (michaelRichards99) [reply tests/test_misc.py]: done.

File: tests/test_misc.py (hunk 7 above)
Comment (michaelRichards99) [reply tests/test_misc.py]: done.

File: tests/test_misc.py (hunk 6 above)
Comment (michaelRichards99) [reply tests/test_misc.py]: It's tricky to directly mock python built-ins such as as DateTime without adding a new 3rd party library or manually mocking with some sort of reflection. I've mocked _get_timestamp from the OutputManager class instead.

File: RUFAS/output_manager.py (hunk 5 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: done.

File: RUFAS/output_manager.py (hunk 4 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: I copied the type from another method in the OutputManager. I removed the method completely so no longer applies.

File: RUFAS/output_manager.py (hunk 3 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: method removed.

File: RUFAS/output_manager.py (hunk 2 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: done.

File: RUFAS/output_manager.py (hunk 1 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py:217]: done.

**Review by PooyaHekmati — APPROVED** [review]
[LGTM](https://lgtm.lol/i/718)

File: RUFAS/output_manager.py [hunk 10]
```diff
@@ -187,9 +189,25 @@ def add_error(self, name: str, msg: str, info_map: Dict[str, Any]) -> None:
...
         """
+        info_map["timestamp"] = self._get_timestamp(include_millis=True)
         key = self._generate_key(name, info_map)
         self._add_to_pool(self.errors_pool, key, msg, info_map)
 
+    def _get_timestamp(self, include_millis: bool = False) -> str:
+        """
+        Produces the current system time as a timestamp string.
+
+        Example
```
Comment (Isgotkowitz) [inline RUFAS/output_manager.py]: There should be `Parameters` and `Returns` sections here

File: RUFAS/output_manager.py [hunk 11]
```diff
@@ -187,9 +189,25 @@ def add_error(self, name: str, msg: str, info_map: Dict[str, Any]) -> None:
...
+        ------
+            >>> self._get_timestamp(include_millis=True)
+            28-Jun-2023_Wed_15-48-21.406585
+            >>> self._get_timestamp(include_millis=False)
+            28-Jun-2023_Wed_15-48-21
```
Comment (Isgotkowitz) [inline RUFAS/output_manager.py]:
```suggestion
        Examples
        --------
        >>> self._get_timestamp(include_millis=True)
        28-Jun-2023_Wed_15-48-21.406585
        >>> self._get_timestamp(include_millis=False)
        28-Jun-2023_Wed_15-48-21
```

**Review by Isgotkowitz — CHANGES_REQUESTED** [review]
Nice PR Michael, looks very good overall. First comment is on some missing sections in the `_get_timestamp()` docstring, and the second has some minor formatting corrections to the `Examples` section of the same docstring.

File: RUFAS/output_manager.py (hunk 11 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: Done.

File: RUFAS/output_manager.py (hunk 10 above)
Comment (michaelRichards99) [reply RUFAS/output_manager.py]: Done.

**Review by Isgotkowitz — APPROVED** [review]
Nice

**Pull request merged**